### PR TITLE
https://github.com/DEX-Company/starfish-java/issues/64

### PR DESCRIPTION
**Pull request by AyushVerma2**
https://github.com/DEX-Company/starfish-java/issues/64

### CODE REVIEW

File: src/main/java/sg/dex/starfish/impl/memory/AMemoryOperation.java [hunk 1]
```diff
@@ -19,7 +23,33 @@
...
+        if(meta.get(Constant.PROVENANCE)== null) {
+            // adding default provenance
+            String actId = UUID.randomUUID().toString();
+            // check if DID is present in agent then agent id will be did
+            // else will create a random UUID
+            String agentId = memoryAgent.getDID()!=null?
+                    memoryAgent.getDID().toString():
+                    UUID.randomUUID().toString();
+
+            Map<String, Object> provenanceData = ProvUtil.createPublishProvenance(actId, agentId);
```
Comment (shark8me) [inline src/main/java/sg/dex/starfish/impl/memory/AMemoryOperation.java]: For operations, we need to use `createInvokeProvenance`, not `createPublishProvenance`. The first has more arguments (e.g. list of asset dependencies.)

**Review by shark8me — CHANGES_REQUESTED** [review]
Needs to use Invoke-specific provenance

File: src/main/java/sg/dex/starfish/impl/memory/AMemoryOperation.java [hunk 2]
```diff
@@ -19,7 +23,33 @@
 public abstract class AMemoryOperation extends AMemoryAsset implements Operation {
 
     protected AMemoryOperation(String metaString, MemoryAgent memoryAgent) {
-        super(metaString, memoryAgent);
+        super(getMetaWithProv(metaString,memoryAgent), memoryAgent);
```
Comment (mikera) [inline src/main/java/sg/dex/starfish/impl/memory/AMemoryOperation.java]: Don't create provenance when a metadata String is provided directly. Only do it when default metadata is being created. 

Please write tests to verify this.

File: src/main/java/sg/dex/starfish/impl/memory/AMemoryOperation.java [hunk 3]
```diff
@@ -19,7 +23,33 @@
...
+        //check if already provenance is present
+
+        if(meta.get(Constant.PROVENANCE)== null) {
+            // adding default provenance
+            String actId = UUID.randomUUID().toString();
+            // check if DID is present in agent then agent id will be did
+            // else will create a random UUID
+            String agentId = memoryAgent.getDID()!=null?
+                    memoryAgent.getDID().toString():
+                    UUID.randomUUID().toString();
```
Comment (mikera) [inline src/main/java/sg/dex/starfish/impl/memory/AMemoryOperation.java]: Definitely should not be creating random UUIDs here!

@shark8me please ensure that the provenance DEP specifies what happens by default (probably a "should" or "may")

File: src/main/java/sg/dex/starfish/impl/memory/AMemoryOperation.java [hunk 4]
```diff
@@ -19,7 +23,33 @@
...
+     * @param metaString Metadata as string
+     * @param memoryAgent agent
+     * @return metadata with PROVENANCE
+     */
+    private static String getMetaWithProv(String metaString,MemoryAgent memoryAgent) {
+
+        Map<String,Object> meta = JSON.toMap(metaString);
+        //check if already provenance is present
+
+        if(meta.get(Constant.PROVENANCE)== null) {
```
Comment (mikera) [inline src/main/java/sg/dex/starfish/impl/memory/AMemoryOperation.java]: What is the logic here? Please add comments describing why something is being done. Ideally reference DEP12

File: src/test/java/sg/dex/starfish/impl/operations/TestMemoryOperations.java [hunk 5]
```diff
@@ -81,6 +86,11 @@ public void testPrimeAsync()  {
         test.put("input", "10");
 
         Job job = memoryOperation.invokeAsync(test);
+        String provenance= memoryOperation.getMetadata().get(Constant.PROVENANCE).toString();
+
+        Map<String,Object> provMap=JSON.toMap(provenance);
+        assertTrue(provMap.get("agent").toString().contains("opf:did:"));
```
Comment (mikera) [inline src/test/java/sg/dex/starfish/impl/operations/TestMemoryOperations.java]: Please add comments to describe what is being tested. @shark8me please work with Ayush to ensure these are consistent with DEP12

File: src/test/java/sg/dex/starfish/impl/operations/TestMemoryOperations.java [hunk 6]
```diff
@@ -154,6 +176,10 @@ public void testHashAsyncFailed()  {
         } catch (Exception e) {
             /* OK, Expected */
         }
+        String provenance= epicFailOperation.getMetadata().get(Constant.PROVENANCE).toString();
```
Comment (mikera) [inline src/test/java/sg/dex/starfish/impl/operations/TestMemoryOperations.java]: I think a Asset.getProvenance() method is appropriate as a shortcut for this. Will be a pretty important method.

File: src/test/java/sg/dex/starfish/impl/operations/TestMemoryOperations.java [hunk 7]
```diff
@@ -154,6 +176,10 @@ public void testHashAsyncFailed()  {
...
+
+        Map<String,Object> provMap=JSON.toMap(provenance);
+        assertTrue(provMap.get("agent").toString().contains("opf:did:"));
```
Comment (mikera) [inline src/test/java/sg/dex/starfish/impl/operations/TestMemoryOperations.java]: Need to do a better test than this. Exactly what DID are we expecting?

**Review by mikera — CHANGES_REQUESTED** [review]

File: src/main/java/sg/dex/starfish/impl/memory/AMemoryOperation.java (hunk 3 above)
Comment (shark8me) [reply src/main/java/sg/dex/starfish/impl/memory/AMemoryOperation.java]: [DEP12](https://github.com/DEX-Company/DEPs/tree/master/12#agents-1) does specify what the agent id should be

**Review by AyushVerma2 — COMMENTED** [review]
looks good!

**Pull request merged**